uk.co.compucorp.civicrm.hrcore/CRM/HRCore/Service/CiviHRStatsGatherer.php

```diff
@@ -0,0 +1,172 @@
+use CRM_HRCore_CMSData_Variable_DrupalVariableService as DrupalVariableService;
+
+/**
+ * Responsible for gathering all required site statistics that will be sent to
+ * monitor site usage.
+ */
+class CRM_HRCore_Service_CiviHRStatsGatherer {
+
+  /**
+   * @var VariableServiceInterface
+   */
+  protected $cmsVariableService;
+
+  /**
+   * @param VariableServiceInterface $cmsVariableService
+   */
+  public function __construct(VariableServiceInterface $cmsVariableService) {
+    $this->cmsVariableService = $cmsVariableService;
+  }
+
+  /**
+   * Fetch and set all required statistics.
+   *
+   * @return CiviHRStatistics
+   */
+  public function gather() {
+    $stats = new CiviHRStatistics();
+    $stats->setGenerationDate(new \DateTime());
+    $stats->setSiteName($this->cmsVariableService->get('site_name'));
+    $this->setBaseUrl($stats);
+    $this->setEntityCounts($stats);
+    $this->setContactSubtypes($stats);
+    $this->setReportConfigurations($stats);
+    $this->setAgeGroups($stats);
+
+    return $stats;
+  }
+
+  /**
+   * Sets the site base URL
+   *
+   * @param CiviHRStatistics $stats
+   */
+  private function setBaseUrl(CiviHRStatistics $stats) {
+    $config = & CRM_Core_Config::singleton();
+    $stats->setSiteUrl($config->userFrameworkBaseURL);
+  }
+
+  /**
+   * Fetches counts and sets them for all required entities
+   *
+   * @param CiviHRStatistics $stats
+   * @throws CiviCRM_API3_Exception
+   */
+  private function setEntityCounts(CiviHRStatistics $stats) {
+    $taskAssignmentsKey = 'uk.co.compucorp.civicrm.tasksassignments';
+    $leaveAndAbsenceKey = 'uk.co.compucorp.civicrm.hrleaveandabsences';
+    $recruitmentKey = 'org.civicrm.hrrecruitment';
+
+    $stats->setEntityCount('contact', $this->getEntityCount('Contact'));
+
+    if (ExtensionHelper::isExtensionEnabled($taskAssignmentsKey)) {
+      $stats->setEntityCount('assignment', $this->getEntityCount('Assignment'));
+      $stats->setEntityCount('task', $this->getEntityCount('Task'));
+      $stats->setEntityCount('document', $this->getEntityCount('Document'));
+    }
+
+    if (ExtensionHelper::isExtensionEnabled($leaveAndAbsenceKey)) {
+      // leave requests in last 100 days
+      $format = 'Y-m-d H:i:s';
+      $oneHundredDaysAgo = (new \DateTime('today - 100 days'))->format($format);
+      $params = ['from_date' => ['>=' => $oneHundredDaysAgo]];
+      $last100DaysCount = (int) civicrm_api3('LeaveRequest', 'getcount', $params);
+      $stats->setEntityCount('leaveRequestInLast100Days', $last100DaysCount);
+
+      // total leave requests
+      $leaveRequestCount = $this->getEntityCount('LeaveRequest');
+      $stats->setEntityCount('leaveRequest', $leaveRequestCount);
+    }
+
+    if (ExtensionHelper::isExtensionEnabled($recruitmentKey)) {
+      $stats->setEntityCount('vacancy', $this->getEntityCount('HRVacancy'));
+    }
+
+    $stats->setEntityCount('drupalUser', $this->getEntityCount('UFMatch'));
```

Comment:
A few weeks ago I was doing a few tests and I noticed that there are some situations where UFMatch can get out of sync. Can you please check what happens in these cases:
- Disable a use in Drupal
- Delete a user in Drupal (will the respective UFMatch still exist?)
- Delete a Contact in Civi  (will the respective UFMatch still exist?)
- Permanently delete a contact
- Manually update UFMatch using the API to make the same user match two or more different contacts
- Manually update UFMatch using the API to make the same contact match two or more users

Reply:
Ah, are you trying to get me to do your dirty work? :smile: 

This ticket is already going over the estimated time, I can do this after if you want, please create a ticket and assign it to me.

Reply:
Not really. What I'm asking is if you confirmed the UFMatch API will always return the information that you're expecting. This is the class responsible to gather that information, so it's important that it gets it from the right place. I'm just suggesting a few scenarios where I think you'll have problems.

Reply:
1. Disable a use in Drupal => UF_MATCH & CIVI CONTACT ARE NOT AFFECTED
1. Delete a user in Drupal (will the respective UFMatch still exist?)  => NO, JUST CIVI CONTACT
1. Delete a Contact in Civi (will the respective UFMatch still exist?)  => NO, JUST DRUPAL USER UNTIL NEXT LOGIN
1. Permanently delete a contact (will the respective UFMatch still exist?)  => NO, JUST DRUPAL USER UNTIL NEXT LOGIN
1. Delete and restore a contact (will the UFMatch be recreated?)  => NO, NOT UNTIL NEXT LOGIN OR MANUAL SYNC
1. Manually update UFMatch using the API to make the same user match two or more different contacts  => UFMATCH POINTING TO SAME DRUPAL USER IS CREATED, CANNOT LOGIN USING THE EMAIL FROM THIS UFMATCH
1. Manually update UFMatch using the API to make the same contact match two or more users  => DOES NOT WORK, UNIQUE KEY ON DOMAIN + CONTACT ID

Some of these points could affect the user count to produce an inaccurate count, I left a comment in the ticket to Mike about these problems. If they will be a big problem we can fix them in another technical task I think.